Reformat C and Python source code using uncrustify and black

dpgeorge opened the pull request:
This PR combines #5691 (uncrustify for C) and #5699 (black for Python) to reformat the majority of the code in this repository.  See discussion in #4223 for further background.

The script `tools/codeformat.py` does the formatting, and run without args will format everything.  Run with args it will format just the given files.

I'm pretty happy with this now, IMO it strikes a good balance between keeping existing style and not changing too many lines, and enforcing a relatively strict style (especially for Python code).

Thanks to @jimmo for helping with `codeformat.py` and fine tuning uncrustify options.

----

I intend to merge this very soon.  If anyone has any comments then please make them ASAP so they can be addressed.

FYI @stinos @Jongy @tannewt @MrSurly @codefreax @dlech @dhalbert

dpgeorge committed:
tools/codeformat.py: Add formatter using uncrustify for C, black for Py.

This commit adds a tool, codeformat.py, which will reformat C and Python
code to fit a certain style.  By default the tool will reformat (almost)
all the original (ie not 3rd-party) .c, .h and .py files in this
repository.  Passing filenames on the command-line to codeformat.py will
reformat only those.  Reformatting is done in-place.

uncrustify is used for C reformatting, which is available for many
platforms and can be easily built from source, see
https://github.com/uncrustify/uncrustify.  The configuration for uncrustify
is also added in this commit and values are chosen to best match the
existing code style.  A small post-processing stage on .c and .h files is
done by codeformat.py (after running uncrustify) to fix up some minor
items:
- space inserted after * when used as multiplication with sizeof
- #if/ifdef/ifndef/elif/else/endif are dedented by one level when they are
  configuring if-blocks and case-blocks.

For Python code, the formatter used is black, which can be pip-installed;
see https://github.com/psf/black.  The defaults are used, except for line-
length which is set at 99 characters to match the "about 100" line-length
limit used in C code.

The formatting tools used and their configuration were chosen to strike a
balance between keeping existing style and not changing too many lines of
code, and enforcing a relatively strict style (especially for Python code).
This should help to keep the code consistent across everything, and reduce
cognitive load when writing new code to match the style.

dpgeorge committed:
py: Removing dangling "else" to improve code format consistency.

dpgeorge committed:
all: Add *FORMAT-OFF* in various places.

This string is recognised by uncrustify, to disable formatting in the
region marked by these comments.  This is necessary in the qstrdef*.h files
to prevent modification of the strings within the Q(...).  In other places
it is used to prevent excessive reformatting that would make the code less
readable.

dpgeorge committed:
py/bc0.h: Shift comment to start of line to improve format consistency.

dpgeorge committed:
py/builtinimport: Adjust if-block order in find_file to clean up #if's.

dpgeorge committed:
unix/unix_mphal: Adjust #if in mp_hal_stdin_rx_chr to improve format.

stinos commented:
Looks good.

Not really used to the style where black puts ): on it's own non-indented line for long argument lists but it's just a matter of time before that becomes readable :)

One thing maybe: I know this is a semi-religuous discussion but my opinion is that it's not the eighties anymore and unless you're developing on a smartphone I think most people have the screen estate to use more than 88 characters on a line. I might be wrong but increasing that a bit shouldn't affect people (?) but would lead to more tgings fitting one 1 line.

Running codeformat.py on Windows just exits with `The command line is too long.` which I think is a limitation when using `os.system`. I'll look for a workaround using `subprocess`.

stinos commented on tools/codeformat.py:
```diff
@@ -0,0 +1,117 @@
+
+    # Setting only one of -c or -p disables the other. If both or neither are set, then do both.
+    format_c = args.c or not args.p
+    format_py = args.p or not args.c
```
forgot to use these below

jimmo replied:
Oh, this was my fault. The original version invoked uncrustify/black for each file individually, and rewriting it to run them all at once I forgot to use these vars. (I changed this because it dramatically speeds things up, partly due to process creation overhead, especially for black, but mostly also parsing config etc).

But speaking of that -- I wondered if there might be a problem with maximum length as you mentioned. Rather than messing with subprocess, it might just be easier to batch them into N files at a time.

stinos replied:
As far as I can tell limit is 32767 on windows (7 and beyond). With all c files I'm around 4100 so that fails. On linux it's ARG_MAX which is normally much more, but still might cause problems. So yeah: batching would be the way to go.

dpgeorge replied:
fixed

stinos commented on tools/codeformat.py:
```diff
@@ -0,0 +1,117 @@
+import argparse
+import glob
+import os
+import sys
```
unused import

dpgeorge replied:
fixed

stinos commented on tools/codeformat.py:
```diff
@@ -0,0 +1,117 @@
+
+    # Format C files with uncrustify.
+    os.system(
+        "uncrustify -c '{}' -l C --no-backup {}".format(
```
if there are no files uncrustify waits for standard input so blocks the script, would be better to skip the call then

jimmo replied:
Batching would fix this too.

dpgeorge replied:
should be fixed

stinos commented:
So I have Uncrustify-0.70.1_f which I think is the latest but it says
```
uncrustify.cfg:609: unknown option 'sp_before_oc_proto_list'
uncrustify.cfg:1228: unknown option 'indent_off_after_assign'
uncrustify.cfg:1237: unknown option 'indent_oc_inside_msg_sel'
uncrustify.cfg:1311: unknown option 'indent_inside_ternary_operator'
uncrustify.cfg:1314: unknown option 'indent_off_after_return'
uncrustify.cfg:1758: unknown option 'nl_func_call_args_multi_line_ignore_closures'
```
which results is exit code 1

stinos commented on tools/codeformat.py:
```diff
@@ -0,0 +1,117 @@
+    def lang_files(exts):
+        for file in files:
+            if os.path.splitext(file)[1].lower() in exts:
+                yield "'" + file + "'"
```
Seems this doesn't work on windows: files need to be surrounded by double quotes. Same for the uncrustify.cfg below.

dpgeorge replied:
fixed

stinos replied:
ok works

stinos commented on tools/codeformat.py:
```diff
@@ -0,0 +1,117 @@
+                yield "'" + file + "'"
+
+    # Format C files with uncrustify.
+    os.system(
```
replace this with subprocess.check_output to overcome 260 (or so) limit on windows

dpgeorge replied:
os.system is still used but the files are batched (eg 200 at a time) to keep the line length down.

stinos replied:
200 is still way to long when using os.system though

stinos replied:
But when using like 20, having to start the process that many times makes things slower..

dpgeorge replied:
ok, now using subprocess.check_call, but I removed the double quotes from the filenames so please double check it still works (and I'll add the quotes back if needed).

dpgeorge commented:
> Can others check what version they have and see if there's a diff?

I'm using `Uncrustify_d-0.70.1-223-fd5e1ddd`, from github source.  I'll try with your version and see the diff.

dpgeorge commented:
> So I have Uncrustify-0.70.1_f which I think is the latest but it says

I tested this version and indeed it gives the diff above when running on this PR.

So.... which one to use?  The latest version is better, the 0.70.1 is arguably buggy in this respect because it's not putting spaces around the multiplication.

stinos commented:
Yeah that's buggy indeed so should be using the latest then, but it's a bit tricky to tell users 'build uncrustify from source then format your code then make a PR'. And if you don't you have to format yourself before merging. But maybe that has to be done anyway, so I'd just go with the version from source.

tannewt commented:
Has there been any work on adding an `uncrustify` check to the CI?

codefreax commented on py/binary.c:
```diff
@@ -297,8 +297,9 @@ void mp_binary_set_val(char struct_type, char val_type, mp_obj_t val_in, byte *p
+            }
             #endif
+
             {
```
The braces (l 303, 313) can now be removed and the code outdented.

dpgeorge replied:
fixed

codefreax commented on extmod/modbluetooth.c:
```diff
@@ -852,7 +853,7 @@ STATIC mp_obj_t bluetooth_ble_invoke_irq(mp_obj_t none_in) {
             ringbuf_extract(&o->ringbuf, data_tuple, 2, 0, NULL, 0, 0, NULL, NULL);
-        #if MICROPY_PY_BLUETOOTH_ENABLE_CENTRAL_MODE
+            #if MICROPY_PY_BLUETOOTH_ENABLE_CENTRAL_MODE
         } else if (event == MP_BLUETOOTH_IRQ_SCAN_RESULT) {
```
Better a few issues like this than outdenting all PP directives?

dpgeorge replied:
now fixed

codefreax commented on extmod/modframebuf.c:
```diff
@@ -491,7 +497,7 @@ STATIC mp_obj_t framebuf_blit(size_t n_args, const mp_obj_t *args) {
+        ) {
         // Out of bounds, no-op.
         return mp_const_none;
     }
```
Readability before was better.

dpgeorge replied:
there might be some uncrustify options that improve this, but for now I think we just leave these few cases as-is, maybe fix them later on

codefreax commented on extmod/modlwip.c:
```diff
@@ -830,14 +832,15 @@ STATIC mp_obj_t lwip_socket_make_new(const mp_obj_type_t *type, size_t n_args, s
             break;
-        #if MICROPY_PY_LWIP_SOCK_RAW
+            #if MICROPY_PY_LWIP_SOCK_RAW
         case MOD_NETWORK_SOCK_RAW: {
```
...

dpgeorge replied:
fixed

codefreax reviewed (commented):
Mainly the preprocessor directive indentation leaves room for improvement IMHO. Overall, this looks great.

Jongy commented on tools/codeformat.py:
```diff
@@ -0,0 +1,127 @@
+            line = " ".join(itertools.islice(files, N))
+            if not line:
+                break
+            os.system(cmd + " " + line)
```
Any particular reason for using `os.system` instead of `subprocess.check_call`? Not referring to the arguments limit, but to the unchecked exit code.

dpgeorge replied:
now using `subprocess.check_call`

Jongy commented on tools/codeformat.py:
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
```
Commit message says `codeformat.sh`, not `.py` :)

dpgeorge replied:
fixed!

Jongy commented on py/gc.c:
```diff
@@ -473,10 +473,38 @@ void *gc_alloc(size_t n_bytes, unsigned int alloc_flags) {
-            if (ATB_1_IS_FREE(a)) { if (++n_free >= n_blocks) { i = i * BLOCKS_PER_ATB + 1; goto found; } } else { n_free = 0; }
-            if (ATB_2_IS_FREE(a)) { if (++n_free >= n_blocks) { i = i * BLOCKS_PER_ATB + 2; goto found; } } else { n_free = 0; }
-            if (ATB_3_IS_FREE(a)) { if (++n_free >= n_blocks) { i = i * BLOCKS_PER_ATB + 3; goto found; } } else { n_free = 0; }
+            if (ATB_0_IS_FREE(a)) {
```
It was better off previously. Specific places like here could use a `// noqa` or `// NOLINT` or whatever.

dpgeorge replied:
fixed, added comments to disable formatting on this part

Jongy commented:
I got the same diff as @stinos with the same uncrustify version on Arch Linux. If it's possible to pin the version, it's probably a good idea.

Overall it looks great. Even if I don't praise much of the style, I think that any enforced style is for the best, as long as its not absolutely awful... I'd be happy to see this merged :+1:

dpgeorge committed:
py/malloc: Put { on separate line for funcs that have selective sigs.

To ensure there are balanced {}'s in the file, and to help with formatting.

dpgeorge committed:
py: Un-nest configuration #if/#endif's for selection of complex code.

Because un-nested #if's are simpler to handle with formatting tools.

dpgeorge committed:
examples/accellog.py: Shift long comments to their own line.

To improve interaction with black formatter.

dpgeorge committed:
examples/bluetooth/ble_temperature_central.py: Shorten comment.

So the line length is less than 100 characters.

dpgeorge committed:
py/parse: Add parenthesis around calculated bit-width in struct.

To improve interaction with uncrustify formatter.

dpgeorge committed:
all: Reformat C and Python source code with tools/codeformat.py.

This is run with uncrustify 0.70.1, and black 19.10b0.

dpgeorge committed:
travis: Add CI job to check code formatting.

The CI job will fail if there is any code which does not conform to the
style encoded by tools/codeformat.py.  And it will list any changes
required.

uncrustify is built from source because Ubuntu bionic has uncrustify-0.66.1
which is from 2017/11/22 and is missing many options needed here.

dpgeorge commented:
> Not really used to the style where black puts ): on it's own non-indented line for long argument lists but it's just a matter of time before that becomes readable :)

Yeah, a bit foreign.



> One thing maybe: I know this is a semi-religuous discussion but my opinion is that it's not the eighties anymore and unless you're developing on a smartphone I think most people have the screen estate to use more than 88 characters on a line. I might be wrong but increasing that a bit shouldn't affect people (?) but would lead to more tgings fitting one 1 line.

Ok, I've now increased the line-length that black's using to 99 characters.  This does lead to a smaller diff in this PR, and it matches the "about 100" line-length limit of the C code.

dpgeorge commented:
> Yeah that's buggy indeed so should be using the latest then, but it's a bit tricky to tell users 'build uncrustify from source then format your code then make a PR'. And if you don't you have to format yourself before merging. But maybe that has to be done anyway, so I'd just go with the version from source.

I've fixed this by adding a small post-processing step to the `codeformat.py` script, which uses some simple regex's to fix what uncrustify got wrong.  This step also fixes up the indenting of #if's in a few places to make it much more consistent with the old styling, and leads to a smaller diff.

dpgeorge commented:
> Has there been any work on adding an `uncrustify` check to the CI?

I just added a commit that runs `codeformat.py` on Travis and then does a `git diff --exit-code`.  So it'll report any differences and fail CI if there are some.

dpgeorge commented:
> Mainly the preprocessor directive indentation leaves room for improvement IMHO

This is now fixed with a simple post-processing stage in `codeformat.py`.

dpgeorge commented on examples/accellog.py:
```diff
@@ -2,16 +2,20 @@
-blue.on()                                           # turn on blue LED
+log = open(
+    "/sd/log.csv", "w"
+)  # open file to write data - /sd/ is the SD-card, /flash/ the internal memory
```
moved

dpgeorge commented on examples/accellog.py:
```diff
@@ -2,16 +2,20 @@
-        log.write('{},{},{},{}\n'.format(t,x,y,z))  # write data to file
+for i in range(
+    100
+):  # do 100 times (if the board is connected via USB, you can't write longer because the PC tries to open the filesystem which messes up your file.)
```
moved

dpgeorge commented on examples/bluetooth/ble_temperature_central.py:
```diff
@@ -72,8 +80,10 @@ def _irq(self, event, data):
-                self._name = decode_name(adv_data) or '?'
+                self._addr = bytes(
+                    addr
+                )  # Note: The addr buffer is owned by modbluetooth, need to copy it.
```
comment shortened to fit on the line

dpgeorge commented on examples/conwaylife.py:
```diff
@@ -1,46 +1,53 @@
-                lcd.pixel(x, y, 1)   # exactly 3 neighbours around an empty cell: cell is born
+                lcd.pixel(
+                    x, y, 1
+                )  # exactly 3 neighbours around an empty cell: cell is born
```
now fits in new line-limit of 99 chars

dpgeorge commented:
> I got the same diff as @stinos with the same uncrustify version on Arch Linux. If it's possible to pin the version, it's probably a good idea.

This should be now fixed, I checked both v0.70.1 and latest master of uncrustify and they now give the same result.

dpgeorge commented:
I've also changed the comment used to disable uncrustify formatting, from `*INDENT-OFF*` to `*FORMAT-OFF*`, because I think it makes it more readable with `*FORMAT-OFF*`.

dpgeorge commented:
For the CI looks like uncrustify will need to be build from source, because Ubuntu 18 includes uncrustify-0.66.1 which is from 2017/11/22 and is missing many options that are used here.

stinos commented:
All looking pretty good now.

dpgeorge commented:
Ok, this PR now passes CI with the code-format check.  (Coverage percentage is decreased most likely because the overall line count has changed, not because there's new uncovered code.)

dpgeorge commented:
Merged!  Thanks to all involved for discussion, review, suggestions, etc.

tve commented:
It would be nice for codeformat.py to check for the existence of uncrusify and provide a link for installation if not present (and talk about version required). Same for black.
